File: client/src/main/java/com/jraska/github/client/RealDeepLinkLauncher.java
package com.jraska.github.client;

import com.jraska.github.client.ui.BaseActivity;
import com.jraska.github.client.ui.UserDetailActivity;
import com.jraska.github.client.ui.UsersActivity;
import okhttp3.HttpUrl;

import javax.inject.Provider;

final class RealDeepLinkLauncher implements DeepLinkLauncher {
  private final Provider<BaseActivity> topActivityProvider;

  RealDeepLinkLauncher(Provider<BaseActivity> topActivityProvider) {
    this.topActivityProvider = topActivityProvider;
  }

  @Override public void launch(HttpUrl deepLink) {
    if (!deepLink.host().equals("github.com")) {
      throw new IllegalArgumentException("We handle only GitHub deep links, not: " + deepLink);
    }

    if ("/users".equals(deepLink.encodedPath())) {
      UsersActivity.start(topActivityProvider.get());
      return;
    }

    if (deepLink.pathSize() == 1) {
      String login = deepLink.pathSegments().get(0);
      UserDetailActivity.start(topActivityProvider.get(), login);
      return;
    }

    throw new IllegalArgumentException("Unexpected deep link: " + deepLink);
  }
}

File: client/src/main/java/com/jraska/github/client/RealDeepLinkLauncher.kt
package com.jraska.github.client

import com.jraska.github.client.ui.BaseActivity
import com.jraska.github.client.ui.RepoDetailActivity
import com.jraska.github.client.ui.UserDetailActivity
import com.jraska.github.client.ui.UsersActivity
import okhttp3.HttpUrl
import timber.log.Timber
import javax.inject.Provider

class RealDeepLinkLauncher private constructor(
  private val topActivityProvider: Provider<BaseActivity>,
  private val launchers: List<LinkLauncher>
) : DeepLinkLauncher {
  override fun launch(deepLink: HttpUrl) {
    if (deepLink.host() != "github.com") {
      throw IllegalArgumentException("We handle only GitHub deep links, not: $deepLink")
    }

    Timber.i("Launching %s", deepLink)

    val activity = topActivityProvider.get()
    for (launcher in launchers) {
      val result = launcher.launch(activity, deepLink)
      if (result == LinkLauncher.Result.LAUNCHED) {
        return
      }
    }

    throw IllegalArgumentException("Unexpected deep link: $deepLink")
  }

  companion object {
    fun create(activityProvider: Provider<BaseActivity>, launchers: Set<LinkLauncher>): RealDeepLinkLauncher {
      val sortedLaunchers = launchers.sortedBy { it.priority().value }
      return RealDeepLinkLauncher(activityProvider, sortedLaunchers)
    }
  }
}
